Create separate panels for multiple status messages of different colours #4147

kennho opened the pull request:
Fixes #4147

damithc commented:
You may want to hold back the changes to html files until the other changes are reviewed. Otherwise the diff becomes too big to review in GitHub.

LowWeiLin commented on src/main/java/teammates/common/util/Const.java:
```diff
@@ -521,6 +521,7 @@
         public static final String COMMENTS = "comments";
         public static final String TEAMMATES = "teammates";
     
+        public static final String STATUS_MESSAGES_LIST = "statusMessagesToUser";
```
Are the constants STATUS_MESSAGE and STATUS_MESSAGE_COLOR still used? Look through and remove them from the code if they are not.

LowWeiLin commented on src/main/java/teammates/common/util/StatusMessage.java:
```diff
@@ -1,8 +1,10 @@
 import teammates.common.util.Const.StatusMessageColor;
 
-public class StatusMessage {
+public class StatusMessage implements Serializable {
```
Add java doc style comments for this public class.

LowWeiLin commented on src/main/java/teammates/ui/controller/Action.java:
```diff
@@ -367,21 +367,19 @@ public ActionResult executeAndPostProcess() throws EntityDoesNotExistException {
     }
 
+    /**
+     * Adds the list of status messages generated after performing an action to session.
```
Rephrase this? Adds the list of status messages from ActionResult intto session variables.

LowWeiLin commented on src/main/java/teammates/ui/controller/Action.java:
```diff
@@ -367,21 +367,19 @@ public ActionResult executeAndPostProcess() throws EntityDoesNotExistException {
 
+    /**
+     * Adds the list of status messages generated after performing an action to session.
+     * @param response ActionResult generated from performing a certain Action
```
Just "@param response ActionResult" ?
There should not be any requirement that is it generated by executing an Action, although that is how it is created.

LowWeiLin commented on src/main/java/teammates/ui/controller/AjaxResult.java:
```diff
@@ -36,7 +36,8 @@ public AjaxResult(AccountAttributes account,
     public void send(HttpServletRequest req, HttpServletResponse resp) throws IOException, ServletException {
         
-        req.setAttribute(Const.ParamsNames.ERROR, "" + isError);        
+        req.setAttribute(Const.ParamsNames.ERROR, "" + isError);
```
Const.ParamsNames.ERROR was previously used to colour the status messages red, is it still used?

LowWeiLin commented on src/main/java/teammates/ui/controller/AjaxResult.java:
```diff
@@ -45,18 +46,27 @@ public void send(HttpServletRequest req, HttpServletResponse resp) throws IOExce
         }
+    }
+    
+    private void clearStatusMessageForRequest(HttpServletRequest req) {
```
Is the function clearStatusMessageForRequest needed? addStatusMessagesToPageData is already calling removeAttribute. Maybe addStatusMessagesToPageData can call this function to clear the status message?

LowWeiLin commented on src/main/webapp/js/common.js:
```diff
@@ -83,6 +83,12 @@ var FEEDBACK_RESPONSE_ID = 'responseid';
+var STATUS_INFO = "info";
+var STATUS_WARNING = "warning";
+var STATUS_DANGER = "danger";
+
```
Instead of individual variables, use an enum instead.
There can also be a function to get the class for each enum so that you do not have to use a bunch of if-else statements in setStatusMessage.

LowWeiLin commented on src/main/java/teammates/ui/controller/PageData.java:
```diff
@@ -1015,6 +1017,22 @@ public String getRecipientNames(Set<String> recipients, String courseId, String
+     * Gets the list of status messages.
+     * @return a list of status messages that is to be displayed to the user
+     */
+    public List<StatusMessage> getStatusMessagesToUser() {
```
This function does not seem to be used...

kennho replied:
The getter is needed so that ${data.statusMessagesToUser} can be used at various jsp files.
e.g. <t:statusMessage statusMessagesToUser="${data.statusMessagesToUser}" />

Without the getter I think the data.statusMessagesToUser will give an error.

kennho commented:
Updated with suggested changes @LowWeiLin

LowWeiLin commented on src/main/java/teammates/common/util/StatusMessage.java:
```diff
@@ -1,8 +1,14 @@
+ * The {@code StatusMessage} class encapsulates the text of status message 
+ * and its level of seriousness of the status message (the color of the message).
+ */
+public class StatusMessage implements Serializable {
```
What is the reason for using Serializable here?

kennho replied:
Ohh. It is to allow the StatusMessage to be added to the session. If not it will throw NotSerializableException when I try to add it to the session.

LowWeiLin commented on src/main/webapp/js/common.js:
```diff
@@ -148,6 +159,11 @@ $(document).on('ajaxComplete ready', function() {
         $tooltips.tooltip('disable');
     }
+    
+    // Freezes the StatusType so as to prevent changing the value accidentally.
```
This is nice, but we don't use it for all our constants in js. Can remove this.

kennho commented:
Updated with suggested changes @LowWeiLin

LowWeiLin commented:
Can you check for js errors? I found this one:
![image](https://cloud.githubusercontent.com/assets/3910365/13659744/37b0987e-e6bf-11e5-93ea-54e9866ac070.png)

kennho commented:
Updated the affected javascript @LowWeiLin  I think I have changed all but is there anyway I can check these errors? Because this change affects too many places and it's abit hard to track if there are any remaining ones.

LowWeiLin commented:
Either use static analysis tools or... manual checking is the only way i can think of.

thyageshm commented:
@kennho reminder to complete this soon

kennho commented:
Yup it's completed already. Waiting for review. @thyageshm

thyageshm commented:
please remind @LowWeiLin to review

LowWeiLin commented on src/main/webapp/js/common.js:
```diff
@@ -532,56 +543,61 @@ function scrollToTop(duration) {
- * @param message
- * @param error
+ * @param message the text message to be shown to the user
+ * @param status the type (StatusType.SUCCESS, StatusType.INFO, StatusType.WARNING, StatusType.DANGER) of the message
```
just use @param status type
no need to write down all types as you will need to modify this comment if more status types are added in the future.

LowWeiLin commented:
minor comment, looks okay after changing that.

kennho commented:
updated with suggested changes @LowWeiLin

kennho commented:
Updated a new html file that was added in a recent merge to display the status message the new way as it was still using the old way of displaying. @LowWeiLin

kennho commented:
Actually can I merge this? Since it's just a change in one of the html files for the test. Or do I need to wait for the whole process for it to be merge approved? @LowWeiLin

damithc commented:
Can merge after @LowWeiLin checked the new commits.

kennho commented:
Okay
